docs/create_your_own_brick.rst

```diff
@@ -0,0 +1,371 @@
+
+This tutorial assumes that you are already familiar with
+:doc:`bricks <bricks_overview>` and how to use them from a user point of view.
+
+
+Bricks ingredients and recipe
+-----------------------------
+
+All the bricks in blocks inherit directly or indirectly from the
+:class:`.Brick`. There is already a rich inheritance hierarchy of
+bricks implemented in blocks and thus, you should consider which brick level
+to inherit from. Bear in mind that multiple inheritance is often possible and
+advocated whenever it makes sense.
+
+Here are examples of possible bricks to inherit from:
+
+* :class:`.Sequence`: a sequence of bricks.
+* :class:`.Initializable`: a brick that defines a same initialiation scheme
+  (weights and biases) for all its children.
+* :class:`.Feedforward`: declares an interface for bricks with one input and
+  one output.
+* :class:`.Linear`: a linear transformation with optional bias. Inherits from
+  :class:`.Initializable` and :class:`.Feedforward`.
+* :class:`.BaseRecurrent`: the base class for recurrent bricks. Check the
+  :doc:`tutorial about rnns</rnn>` for more information.
+* many more!
+
+Let's say that you want to create a brick from scracth, simply inheriting
+from :class:`.Brick`, then you should consider overwriting the following
+methods (strictly speaking, all these methods are optional):
+
+* :meth:`.Brick.__init__`: you should pass by argument the attributes of your
+  brick. It is also in this method that you should create the potential
+  "children bricks" that belongs to your brick (in that case, you have to put
+  the children bricks into ``self.children``). The initialiazation of the
+  attributes can be lazy as described later in the tutorial.
+* :meth:`apply`: you need to implement a method that actually
+  implements the operation of the brick, taking as arguments the inputs
+  of the brick and returning its outputs. It can have any name and for simple
+  bricks is often named ``apply``. You should decorate it with the
+  :func:`.application` decorator, as explained in the next section. If you
+  design a recurrent brick, you should instead decorate it with the
+  :func:`.recurrent` decorator as explained in the
+  :doc:`tutorial about rnns</rnn>`.
+* :meth:`.Brick._allocate`: you should implement this method to allocate the
+  shared variables (often representing parameters) of the brick. In blocks,
+  by convention, the built-in bricks allocate their shared variables with nan
+  values and we recommand you to do the same.
+* :meth:`.Brick._initialize`: you should implement this method to initialize
+  the shared variables of your brick. This method is called after the
+  allocation.
+* :meth:`.Brick._push_allocation_config`: you should consider overwriting
+  this method if you want to allocate the children bricks in a specific way.
+* :meth:`.Brick._push_initialization_config`: you should consider
+  overwriting this method if you want to initialize the children bricks in a
```

Comment:
The same rephrasal as for `_push_allocation_config` is proposed. Also, I would elaborate on what `Initializable` does (propagates `weights_init` and `biases_init`).